lib/domain.js

```diff
@@ -48,13 +48,56 @@ Object.defineProperty(process, 'domain', {
   }
 });
 
+const pairing = new Map();
+const asyncHook = async_hooks.createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+
+    if (process.domain !== null && process.domain !== undefined) {
+      // if this operation is created while in a domain, let's mark it
+      pairing.set(asyncId, process.domain);
+      resource.domain = process.domain;
+      if (resource.promise !== undefined &&
+        resource.promise instanceof Promise) {
+        // resource.promise instanceof Promise make sure that the
+        // promise comes from the same context
+        // see https://github.com/nodejs/node/issues/15673
+        resource.promise.domain = process.domain;
+      }
+    }
+  },
+  before(asyncId) {
+
+    if (pairing.has(asyncId)) { // enter domain for this cb
+      pairing.get(asyncId).enter();
+    }
+  },
+  after(asyncId) {
+
+    if (pairing.has(asyncId)) { // exit domain for this cb
+      pairing.get(asyncId).exit();
+    }
+  },
+  destroy(asyncId) {
+
+    pairing.delete(asyncId); // cleaning up
+  }
+});
+
 // It's possible to enter one domain while already inside
 // another one. The stack is each entered domain.
 const stack = [];
 exports._stack = stack;
+process._setupDomainUse(stack);
+
+class Domain extends EventEmitter {
```

Comment:
Technically changing this to ES6 style classes is API breaking. We are likely going to land this in node 9 anyway so it should be a problem. Just a note.